### PR TITLE
feat: Encrypt OAuth tokens with AES-256-GCM

### PR DESCRIPTION
**Pull request by WashizuRyo**
### Summary
Encrypts OAuth access tokens and refresh tokens stored in the database using AES-256-GCM encryption. This addresses the security vulnerability where tokens were previously stored in plaintext, which could expose user accounts if the database were compromised.

### Related Issue
N/A

### Changes
- Add `lib/token-encryption.ts` with `encryptToken()` and `decryptToken()` functions using AES-256-GCM
- Encrypt tokens when storing via OAuth callback (`app/(auth)/auth/callback/[provider]/route.ts`)
- Encrypt tokens when refreshing (`app/(auth)/lib/refresh-oauth-credential.ts`)
- Decrypt tokens when fetching (`services/accounts/oauth-credentials.ts`)
- Add `TOKEN_ENCRYPTION_KEY` to `.env.example`
- Add comprehensive unit tests for encryption/decryption

### Testing
- Unit tests cover encryption/decryption round-trip, backward compatibility with plaintext tokens, tamper detection, unicode handling, and error cases
- Manual verification: OAuth re-authentication stores encrypted tokens (prefixed with `enc:`) and connected services work correctly

### Other Information
- **Backward compatibility**: Existing plaintext tokens are automatically handled - `decryptToken()` returns them as-is. Tokens will be encrypted on next OAuth operation (refresh or re-authentication).
- **Deployment requirement**: Set `TOKEN_ENCRYPTION_KEY` environment variable (generate with `node -e "console.log(require('crypto').randomBytes(32).toString('base64'))"`)

___

### **PR Type**
Enhancement

___

### **Description**
- Implement AES-256-GCM encryption for OAuth tokens at rest

- Encrypt tokens during storage and refresh operations

- Decrypt tokens transparently on retrieval with backward compatibility

- Add comprehensive unit tests covering encryption, tampering, and edge cases

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["OAuth Callback"] -->|encryptToken| B["Encrypted Storage"]
  C["Token Refresh"] -->|encryptToken| B
  B -->|decryptToken| D["Service Usage"]
  E["Plaintext Tokens"] -->|backward compatible| D
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>token-encryption.ts</strong><dd><code>AES-256-GCM token encryption utility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/lib/token-encryption.ts

<ul><li>Implement <code>encryptToken()</code> using AES-256-GCM with random IV and <br>authentication tag<br> <li> Implement <code>decryptToken()</code> with backward compatibility for plaintext <br>tokens<br> <li> Validate encryption key from environment variable (32 bytes base64)<br> <li> Use <code>enc:</code> prefix to identify encrypted tokens</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-cf22b1aefaf8f50ec7d833ac5f7e3972f4c6033b6e5fd0c3bc3af86b783721bf">+63/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>route.ts</strong><dd><code>Encrypt OAuth tokens in callback handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/app/(auth)/auth/callback/[provider]/route.ts

<ul><li>Import <code>encryptToken</code> function<br> <li> Encrypt <code>provider_token</code> before storing as <code>accessToken</code><br> <li> Encrypt <code>provider_refresh_token</code> before storing as <code>refreshToken</code><br> <li> Apply encryption to both initial insert and conflict update operations</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-ade206a4905270cae35efe8c6fdcede31a9d42815049cbfb83e6a583b6667745">+10/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>refresh-oauth-credential.ts</strong><dd><code>Encrypt tokens during OAuth refresh</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/app/(auth)/lib/refresh-oauth-credential.ts

<ul><li>Import <code>encryptToken</code> function<br> <li> Encrypt <code>accessToken</code> parameter before database update<br> <li> Encrypt <code>refreshToken</code> parameter before database update</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-c70d90b7ddb8725a54a07364980cb1e495e290956c3fca957363739ffe4d1166">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>oauth-credentials.ts</strong><dd><code>Decrypt OAuth tokens on retrieval</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/services/accounts/oauth-credentials.ts

<ul><li>Import <code>decryptToken</code> function<br> <li> Decrypt <code>accessToken</code> and <code>refreshToken</code> when retrieving credentials<br> <li> Handle null <code>refreshToken</code> gracefully during decryption</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-b4abae487dce63787c52421be274ef78803026cc23ce3875748cbd3090198ab0">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>token-encryption.test.ts</strong><dd><code>Comprehensive token encryption unit tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/lib/token-encryption.test.ts

<ul><li>Test encryption/decryption round-trip correctness<br> <li> Verify <code>enc:</code> prefix on encrypted tokens<br> <li> Ensure non-deterministic encryption (same plaintext produces different <br>ciphertext)<br> <li> Test backward compatibility with plaintext tokens<br> <li> Test tamper detection and error handling for invalid keys<br> <li> Test edge cases: empty strings, unicode, long tokens</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-16d6cd8c3a5fefe459f5c4dea1fa09076434a9e8e00909e4be505e09c383bce1">+90/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>.env.example</strong><dd><code>Document token encryption key configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/.env.example

<ul><li>Add <code>TOKEN_ENCRYPTION_KEY</code> environment variable documentation<br> <li> Include generation command for secure key creation<br> <li> Add comment explaining AES-256-GCM encryption purpose</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2462/files#diff-47ae2e6d55f9b69da2be79d421a82fc5791d777a3051e6cf01d9a7f3998e0e59">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * OAuth tokens are now encrypted at rest and transparently decrypted when accessed.

* **Chores**
  * Added TOKEN_ENCRYPTION_KEY to environment configuration (example updated) to enable encryption.

* **Tests**
  * Added comprehensive tests covering encryption/decryption, edge cases (empty, unicode, long tokens), tampering detection, and key validation/backward compatibility.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

---

> [!NOTE]
> Encrypts stored OAuth access/refresh tokens with AES-256-GCM and decrypts on retrieval, with tests and env config.
> 
> - **Auth/Security**:
>   - **Token encryption utility**: Add `lib/token-encryption.ts` (AES-256-GCM `encryptToken`/`decryptToken`, `enc:` prefix, key validation).
>   - **OAuth flows**:
>     - Callback handler stores encrypted `accessToken`/`refreshToken` and updates on conflict (`app/(auth)/auth/callback/[provider]/route.ts`).
>     - Refresh path writes encrypted tokens (`app/(auth)/lib/refresh-oauth-credential.ts`).
>     - Retrieval decrypts tokens transparently, handling null/legacy plaintext (`services/accounts/oauth-credentials.ts`).
> - **Tests**: Add comprehensive unit tests in `lib/token-encryption.test.ts` (round-trip, tampering, edge cases, key errors).
> - **Config**: Document `TOKEN_ENCRYPTION_KEY` in `.env.example` with generation command.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 237dc820e2df66b4bf762d9a01ab0c5e899ae505. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by WashizuRyo** [comment]
I'm checking agents review.

File: apps/studio.giselles.ai/services/accounts/oauth-credentials.ts [hunk 1]
```diff
@@ -17,5 +18,11 @@ export async function getOauthCredential(provider: OAuthProvider) {
...
 			),
 		);
-	return result?.oauthCredentials;
+
+	const cred = result.oauthCredentials;
+	return {
+		...cred,
+		accessToken: decryptToken(cred.accessToken),
+		refreshToken: cred.refreshToken ? decryptToken(cred.refreshToken) : null,
+	};
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply apps/studio.giselles.ai/services/accounts/oauth-credentials.ts:31]: Fixed at [f2116ba](https://github.com/giselles-ai/giselle/pull/2462/commits/f2116baf46638ab84a778308976c1add1d0b0ab5).

File: apps/studio.giselles.ai/services/accounts/oauth-credentials.ts [hunk 2]
```diff
@@ -17,5 +18,11 @@ export async function getOauthCredential(provider: OAuthProvider) {
 				eq(oauthCredentials.provider, provider),
 			),
 		);
-	return result?.oauthCredentials;
+
+	const cred = result.oauthCredentials;
+	return {
+		...cred,
+		accessToken: decryptToken(cred.accessToken),
+		refreshToken: cred.refreshToken ? decryptToken(cred.refreshToken) : null,
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply apps/studio.giselles.ai/services/accounts/oauth-credentials.ts:30]: Fixed at [f2116ba](https://github.com/giselles-ai/giselle/pull/2462/commits/f2116baf46638ab84a778308976c1add1d0b0ab5).

File: apps/studio.giselles.ai/services/accounts/oauth-credentials.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply apps/studio.giselles.ai/services/accounts/oauth-credentials.ts:31]: Fixed at [f2116ba](https://github.com/giselles-ai/giselle/pull/2462/commits/f2116baf46638ab84a778308976c1add1d0b0ab5).

**Comment by WashizuRyo** [comment]
I'll do QA testing tomorrow.

File: apps/studio.giselles.ai/lib/token-encryption.test.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		});
+
+		test("throws error when encryption key is wrong length", () => {
+			process.env.TOKEN_ENCRYPTION_KEY = randomBytes(16).toString("base64");
+
+			expect(() => encryptToken("test")).toThrow(
+				"TOKEN_ENCRYPTION_KEY must be 32 bytes",
+			);
+		});
+	});
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply apps/studio.giselles.ai/lib/token-encryption.test.ts:109]: Added test at [6339d5b](https://github.com/giselles-ai/giselle/pull/2462/commits/6339d5bfc1526ff66f3ca8388a53751f05dcc287).

File: apps/studio.giselles.ai/lib/token-encryption.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		return value;
+	}
+
+	const key = getEncryptionKey();
+	const data = Buffer.from(value.slice(ENCRYPTED_PREFIX.length), "base64");
+
+	const iv = data.subarray(0, IV_LENGTH);
+	const authTag = data.subarray(IV_LENGTH, IV_LENGTH + AUTH_TAG_LENGTH);
+	const encrypted = data.subarray(IV_LENGTH + AUTH_TAG_LENGTH);
+
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply apps/studio.giselles.ai/lib/token-encryption.ts:61]: Fixed at [e67c709](https://github.com/giselles-ai/giselle/pull/2462/commits/e67c709aa85dc84c823f6763485acef63fc1ef27).

**Comment by WashizuRyo** [comment]
@shige

I’ve finished QA testing on the local server.
Before merging this PR, please ensure that the required environment variables are set in both the preview and production environments, specifically by setting the `TOKEN_ENCRYPTION_KEY` environment variable generated with the following command:

`node -e "console.log(require('crypto').randomBytes(32).toString('base64'))"`
